snapcraft/services/lifecycle.py

```diff
@@ -36,5 +68,47 @@ def setup(self) -> None:
             base=project.get_effective_base(),
             extra_build_snaps=project.get_extra_build_snaps(),
         )
-
         super().setup()
+
+    @overrides
+    def run(self, step_name: str | None, part_names: list[str] | None = None) -> None:
+        super().run(step_name, part_names)
+
+        enable_manifest = utils.strtobool(os.getenv("SNAPCRAFT_BUILD_INFO", "n"))
+        if enable_manifest:
+            self._generate_manifest()
+
+    def _generate_manifest(self) -> None:
+        """Create and populate the manifest file."""
+        emit.progress("Generating snap manifest...")
+        image_information = os.getenv("SNAPCRAFT_IMAGE_INFO", "{}")
+        primed_stage_packages: set[str] = set()
+
+        project = cast(models.Project, self._project)
+
+        parts = copy.deepcopy(project.parts)
+        for name, part in parts.items():
+            assets = self.get_pull_assets(part_name=name)
+            if assets:
+                part["stage-packages"] = assets.get("stage-packages", []) or []
+            for key in ("stage", "prime", "stage-packages", "build-packages"):
+                part.setdefault(key, [])
+
+            stage_packages = self.get_primed_stage_packages(part_name=name)
+            if stage_packages:
+                primed_stage_packages |= set(stage_packages)
+
+        manifest.write(
+            project,
+            self.prime_dir,
+            arch=utils.get_host_architecture(),
+            parts=parts,
+            start_time=self._start_time,
+            image_information=image_information,
+            primed_stage_packages=list(primed_stage_packages),
+        )
+        emit.progress("Generated snap manifest", permanent=True)
+
+        # Also copy the original snapcraft.yaml
+        if self._project_path:
```

Comment:
shouldn't it be an error if `self._project_path` is None?

Reply:
Since this happens at the end of the lifecycle, project path should never be None at this point. But sure, we should issue an error since we're testing it.